### PR TITLE
Add host-independent GitHub/Gitea/Forgejo support to the github drive

### PR DESCRIPTION
**Pull request by CrazyWolf13**
This adds a `github` drive variant that works against GitHub, GitHub Enterprise Server, Gitea, or Forgejo - anywhere that speaks the same "contents API" GitHub does. `apiBase` is the only thing that differs between them, so one adapter covers all four instead of needing separate providers. I built this because I wanted the same option in [Stylus](https://github.com/openstyles/stylus)'s sync settings, which is why it's split across two PRs: this one is the drive-provider itself, the Stylus-side wiring is a separate PR once this lands. Companion PR: openstyles/stylus#

I have already successfully got this into violentmonkey: https://github.com/violentmonkey/violentmonkey/pull/2672 and the corresponding usync: https://github.com/usync-dev/usync/pull/1

A couple of deliberate scope calls:
- The repository has to already exist - this won't create one for you, it just fails clearly (`code: "ENOREPO"`) if it can't find the repo or the token can't see it.
- `apiBase`, `branch`, and `createMethod` all treat an empty string the same as unset, not just `undefined`. A blank env var or a blank field in a settings UI is an easy trap otherwise - I hit this myself while testing (see the regression tests added for it).

> [!NOTE]
> Full disclosure: I built this with a lot of help from Claude (Anthropic's AI) - most of the implementation was AI-generated, with me steering the design decisions and doing the actual testing. I'm not comfortable claiming I could've written all of this myself at this quality in a reasonable amount of time, and I'd rather say that upfront than have it come out in review.

> [!WARNING]
> Given the above, please review this more carefully than you normally would a human-written PR. I've tested it end-to-end against both a real GitHub repo and a real Gitea instance (see below) and it works for my use case, but I may have missed edge cases an experienced maintainer would catch on sight.

## Testing

- `test/adapter/github.offline.js` (new) - 15 unit tests against a fake `fetch`, no credentials needed:
```
  node --test test/adapter/github.offline.js
```
- I ran the full functional suite (`npm test`) against two real accounts — modifying the db before the first sync, a full sync, sync between two instances, 100 changes, and the cloud-locked-while-syncing case — via the existing `test/adapter/github.js` (now with `GITHUB_API_BASE`/`GITHUB_CREATE_METHOD` env support so it can target a self-hosted instance instead of just github.com):

  **GitHub, `createMethod: put`** - full suite passing

  **Gitea, `createMethod: post`** - same suite passing

Screenshot 2026-09-15 221705

<details>
<summary>Tests ran </summary>

  ```bash
  git clone https://github.com/CrazyWolf13/db-to-cloud.git
  cd db-to-cloud
  git checkout feat/github-drive-host-independence
  npm install
  cp .env.example .env
  ```

Fill in `.env`:
- `GITHUB_ACCESS_TOKEN` - a personal access token with read/write access to a repo named `_db_to_cloud_test` under `GITHUB_OWNER` (create the repo first, this adapter never creates one)
- `GITHUB_OWNER` - your GitHub username or org
- for a self-hosted Gitea/Forgejo/GHES run: uncomment `GITHUB_API_BASE` (its API URL) and set `GITHUB_CREATE_METHOD=post` for Gitea/Forgejo

  ```bash
  npm test
  ```

</details>

Let me know if you'd rather see this structured differently.

Thanks, hope this makes it into stylus too

### CODE REVIEW

File: lib/drive/github.js [hunk 1]
```diff
@@ -3,14 +3,31 @@ const base64 = require("universal-base64");
...
+    throw new Error("owner and repo are required");
+  }
+  // `|| default`, not just the destructuring default, because a caller-side
+  // form field or env var left blank (e.g. `GITHUB_API_BASE=`, or an empty
+  // "branch" input in a settings UI) comes through as "", which skips a
+  // destructuring default (only `undefined` triggers those) — apiBase would
+  // otherwise build a request with no host, and branch an empty ?ref=.
+  apiBase = (apiBase || "https://api.github.com").replace(/\/+$/, "");
+  createMethod = createMethod || "put";
+  branch = branch || "main";
```
Comment (eight04) [inline lib/drive/github.js:29]: Move those type check to L19 and update error message. We should never duplicate the default value.

File: lib/drive/github.js [hunk 2]
```diff
@@ -30,93 +48,143 @@ function createDrive({
       args.headers["User-Agent"] = userAgent;
     }
     if (!args.headers["Accept"]) {
-      args.headers["Accept"] = "application/vnd.github.v3+json";
+      // Plain JSON rather than GitHub's vnd.github.v3+json — Gitea/Forgejo
+      // have no reason to honor a GitHub-specific media type.
+      args.headers["Accept"] = "application/json";
+    }
+    if (token && !args.headers["Authorization"]) {
```
Comment (eight04) [inline lib/drive/github.js:55]: Token authorization should be implemented in `createRequest`. I suggest making `getAccessToken` be able to return auth scheme:
```js
getAccessToken: () => String | {scheme: String, param: String}
```
While scheme defaults to `Bearer`.

File: lib/drive/github.js [hunk 3]
```diff
@@ -30,93 +48,143 @@ function createDrive({
...
+      args.headers["Authorization"] = `token ${token}`;
     }
-    args.path = `https://api.github.com${args.path}`;
+    // put()/delete_() re-read `sha` right before writing to avoid a
+    // stale-sha conflict; a cached read would defeat that.
+    args.cache = "no-store";
```
Comment (eight04) [inline lib/drive/github.js:60]: Please elaborate. Why do you want to disable browser cache?

File: lib/drive/github.js [hunk 4]
```diff
@@ -30,93 +48,143 @@ function createDrive({
...
     return request(args);
   }
-  
+
+  function contentsPath(file) {
+    const encoded = file.split("/").filter(Boolean).map(encodeURIComponent).join("/");
+    return `/repos/${owner}/${repo}/contents${encoded ? `/${encoded}` : ""}`;
+  }
+
+  function contentsPathForRead(file) {
```
Comment (eight04) [inline lib/drive/github.js:70]: It's unclear what `contentsPathForRead` is. I suggest changing `contentsPath` to
```js
contentsPath(file, ref = false) => String
```
Note that it shouldn't add `ref` query if `branch` is falsy.

File: lib/drive/github.js [hunk 5]
```diff
@@ -3,14 +3,31 @@ const base64 = require("universal-base64");
...
 const {createRequest} = require("../request");
 
+// Also covers GHES, Gitea, and Forgejo — they implement the same Contents
+// API, so `apiBase` is enough to target any of them; no per-host branching.
 function createDrive({
   userAgent = "db-to-cloud",
+  apiBase = "https://api.github.com",
   owner,
   repo,
+  branch = "main",
```
Comment (eight04) [inline lib/drive/github.js:13]: `branch` should default to null, representing not using any `?ref` query.

File: lib/drive/github.js [hunk 6]
```diff
@@ -3,14 +3,31 @@ const base64 = require("universal-base64");
...
+  token,
```
Comment (eight04) [inline lib/drive/github.js:14]: Not really sure if the adapter should manage personal token. In webdav we save username and password, so I guess it's fine for now.

File: lib/drive/github.js [hunk 7]
```diff
@@ -30,93 +48,143 @@ function createDrive({
...
+      if (err.code === 404) {
+        if (!await checkRepoExists()) {
+          const notFound = new Error(
+            `Repository "${owner}/${repo}" was not found, or the token does not have access to it. ` +
+              `Create the repository first, then make sure the token can read and write its contents.`
+          );
+          notFound.code = "ENOREPO";
+          throw notFound;
+        }
+        return [];
```
Comment (eight04) [inline lib/drive/github.js:102]: `list` is allowed to throw an 404 error. We don't have to create an empty array here.

File: lib/drive/github.js [hunk 8]
```diff
@@ -30,93 +48,143 @@ function createDrive({
...
-    while (!result) {
+    // GitHub uses PUT for both create and update; Gitea/Forgejo need POST
+    // to create (their PUT requires an existing sha) — see createMethod.
+    // On a genuine 409 (another commit landed between our sha lookup and
+    // our write) retry once with a freshly read sha, per GitHub's own docs.
+    for (let attempt = 0; ; attempt++) {
+      let sha;
+      if (overwrite) {
+        try {
+          const existing = await requestAPI({path: contentsPathForRead(file)});
```
Comment (eight04) [inline lib/drive/github.js:130]: A put action should only modify `shaCache` if the operation succeeded.

When `overwrite=true`, the function should take sha from `shaCache` and throw if it doesn't exist.

File: lib/drive/github.js [hunk 9]
```diff
@@ -30,93 +48,143 @@ function createDrive({
...
-      method: "PUT",
-      path: `/repos/${owner}/${repo}/contents/${file}`,
-      contentType: "application/json",
-      body: JSON.stringify(params)
-    };
-    let retried = false;
-    let result;
-    while (!result) {
+    // GitHub uses PUT for both create and update; Gitea/Forgejo need POST
+    // to create (their PUT requires an existing sha) — see createMethod.
```
Comment (eight04) [inline lib/drive/github.js:123]: Please provide document reference.

Gitea seems to support `put` without sha?
https://gitea.com/api/swagger#/repository/repoUpdateFile

File: lib/drive/github.js [hunk 10]
```diff
@@ -30,93 +48,143 @@ function createDrive({
...
   }
-  
+
   async function delete_(file) {
-    try {
-      let sha = shaCache.get(file);
-      if (!sha) {
-        await get(file);
-        sha = shaCache.get(file);
+    // Same stale-sha race as put() applies here.
```
Comment (eight04) [inline lib/drive/github.js:164]: Revert this. A stale sha should fail the entire sync.

Fetching the file while `sha` is missing is just for debugging so that I can easily delete any file in the cloud.

File: test/adapter/github.offline.js [hunk 11]
```diff
@@ -0,0 +1,250 @@
+// Offline unit tests for the github drive, against a fake fetch — no live
```
Comment (eight04) [inline test/adapter/github.offline.js:1]: Not sure if this unit test make sense. The sync machine should be a black box and the adapter shouldn't expect how it will be used.

If we want to test the connection between adapter and the fetch function, we can simply check if `fetch` is getting the correct arguments.

**Review by eight04 — CHANGES_REQUESTED** [review]
I think there are some misunderstanding about `shaCache`.

File: lib/drive/github.js (hunk 1 above)
Comment (CrazyWolf13) [reply lib/drive/github.js]: createMethod is gone now. 
branch/apiBase stay as body-level x || default 
a destructuring default only fires on undefined, not "", and a blank form field sends ""

File: lib/drive/github.js (hunk 5 above)
Comment (CrazyWolf13) [reply lib/drive/github.js]: default is null now, matches. 
Just done via branch = branch || null in the body, same reason as above.

File: lib/drive/github.js (hunk 8 above)
Comment (CrazyWolf13) [reply lib/drive/github.js]: shaCache-only-on-success and no-live-reGET are done. 
But "throw if missing" I changed: sync engine calls put() on brand-new files before ever populating shaCache (verified in lib/db-to-cloud.js), so throwing broke the first sync :/ 

Now: no cached sha -> try create, fetch+retry once only on real conflict.

File: lib/drive/github.js (hunk 9 above)
Comment (CrazyWolf13) [reply lib/drive/github.js]: adressed, I had it initially this way too on the violentmonkey side, and their author seemingly wrongly reviewed this, I tested it and put without sha seems to be working too, so refactored that.

File: test/adapter/github.offline.js (hunk 11 above)
Comment (CrazyWolf13) [reply test/adapter/github.offline.js]: Tests now only check fetch call shape per input (method/url/headers/body), no assumptions about how the sync engine sequences calls.

File: lib/drive/github.js (hunk 8 above)
Comment (eight04) [reply lib/drive/github.js]: > When `overwrite=true`, the function should take sha from `shaCache` and throw if it doesn't exist.
> throwing broke the first sync :/

I was wrong. The old logic should be good: When `overwrite=true` and sha is available, use the sha.
https://github.com/eight04/db-to-cloud/blob/13b2d9dde0603a0b827e7d1ef1ec8a4a346e1541/lib/drive/github.js#L66-L68
So that the first push will be considered as creating files.

There is no valid situation that sha is outdated during `put`.

Comment (CrazyWolf13) [reply lib/drive/github.js]: Should be resolved

**Comment by CrazyWolf13** [comment]
thanks, all review comments should be adressed and ready for a re-review.

File: test/adapter/github.offline.js [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+// account needed. github.js in this same directory covers the real API and
+// needs GITHUB_ACCESS_TOKEN/GITHUB_OWNER (+ optional GITHUB_API_BASE) in .env.
+//
+// Run with: node --test test/adapter/github.offline.js
+const {test} = require("node:test");
+const assert = require("node:assert/strict");
+
+const createDrive = require("../../lib/drive/github");
+
+function makeFakeFetch(routes) {
```
Comment (eight04) [inline test/adapter/github.offline.js]: This should be simplified by switching to `mock.fn` in `node:test`. We probably should merge #34 first.

File: README.md [hunk 13]
```diff
@@ -337,14 +337,35 @@ If `fetch` is not supplied, use global variable `fetch`.
 
 ```js
 github({
-  getAccessToken: async () => token: String,
+  getAccessToken: async () => token: String | {scheme: String, param: String},
```
Comment (eight04) [inline README.md]: Doc can be improved by extracting `getAccessToken` to a new section before adapters.

**Review by eight04 — APPROVED** [review]
LGTM. Will find some time to test it locally and see if it can pass the test against github at lest. I don't have gitea account.

If it doesn't run, I will merge #34 first to update build tools. Should make it easier to build.

Do you want to update unit test and `getAccessToken` doc? Or we can handle them after merge.

File: README.md (hunk 13 above)
Comment (CrazyWolf13) [reply README.md]: done

File: test/adapter/github.offline.js (hunk 12 above)
Comment (CrazyWolf13) [reply test/adapter/github.offline.js]: Yeah, probably better to merge #34 first, I also pushed a commit to use `mock.fn`

**Comment by CrazyWolf13** [comment]
Great :)

For me both github and gitea passed all tests, for gitea I just shortly spun up a local instance:

```bash
docker run --rm -d --name gitea -p 3000:3000 -p 2222:22 gitea/gitea:latest 
```
